guardian/cdk
# feat!: Remove stage awareness from constructs

This is a fairly big change. It is possibly easier to review commit by commit. Happy to discuss ways to break it down - I'd be keen for the changes to yield a single new release.

## What does this change?


Following discussions in #1049, #1048 and IRL, this PR moves to simplify the codebase by removing explicit stage awareness from these constructs:
  - `GuAutoScalingGroup`
  - `GuCertificate`
  - `GuCname`
  - `GuAlarm`

These constructs can still be made stage aware, however it is now the decision of the caller.

This change impacts the `Ec2App` pattern too, but thankfully not in a dramatic way. The prop `certificateProps` has been renamed to `domain`; this is because it is more explicit.

~❓ https://github.com/guardian/cdk-playground/pull/161 demonstrates this branch on a simple stack. If we didn't rename `certificateProps` to `domain` this change would be a no-op. Shall we _not_ rename the prop to increase backwards compatibility?~ reverted.

### Breaking changes
BREAKING CHANGE: The props to `GuAutoScalingGroup` has changed.

Scaling props in `GuAutoScalingGroup` are now of type `number` and no longer need to be decorated with stage details. The caller can optionally provide a Mapping (via `withStageDependentValue`) if necessary. `GuCertificate`, `GuCname`, and `GuAlarm` see a similar change.

From:

```ts
new GuAutoScalingGroup(stack, "AutoscalingGroup", {
   // other required props
  stageDependentProps: {
    [Stage.CODE]: { minimumInstances: 2 },
    [Stage.PROD]: { minimumInstances: 5 },
  },
});
```

To:

```ts
// Stage agnostic
new GuAutoScalingGroup(stack, "AutoscalingGroup", {
   // other required props
  minimumInstances: 2,
  maximumInstances: 11,
});

// Stage aware
const app = "TestApp";

new GuAutoScalingGroup(stack, "AutoScalingGroup", {
   // other required props
  minimumInstances: stack.withStageDependentValue<number>({
    app,
    variableName: "minInstances",
    stageValues: { [Stage.CODE]: 1, [Stage.PROD]: 3 },
  }),
  maximumInstances: stack.withStageDependentValue<number>({
    app,
    variableName: "maxInstances",
    stageValues: { [Stage.CODE]: 2, [Stage.PROD]: 6 },
  }),
});
```

## How to test



See changes to unit tests.

Also see https://github.com/guardian/cdk-playground/pull/161 which tests this branch.

## How can we measure success?


A simpler codebase!

## Have we considered potential risks?


Is the API being presented actually clearer?

## Checklist

  - ℹ️  NOTE: The `GuEc2App` could benefit from a refactor, namely documenting the props more explicitly. This is out of scope for this PR, however.

[^1]: Consider whether this is something that will mean changes to projects that have already been migrated, or to the CDK CLI tool. If changes are required, consider adding a checklist here and/or linking to related PRs.
[^2]: If you are adding a new construct or pattern, has new documentation been added? If you are amending defaults or changing behaviour, are the existing docs still valid?
